google/cloud/aiplatform/initializer.py

```diff
@@ -14,28 +14,142 @@
+
+
+def validate_region(region: str):
+    """Validates region against supported regions.
+
+    Args:
+        region: region to validate
+    Raises:
+        ValueError if region is not in supported regions.
+    """
+    region = region.lower()
+    if region not in SUPPORTED_REGIONS:
+        raise ValueError(
+            f"Unsupported region for AI Platform, select from {SUPPORTED_REGIONS}"
+        )
 
-from typing import Optional
 
 class Init:
+    """Stores common parameters and options for API calls."""
+
+    def __init__(self):
+        self._project = None
+        self._experiment = None
+        self._location = None
+        self._staging_bucket = None
+        self._credentials = None
+
+    def init(
+        self,
+        *,
+        project: Optional[str] = None,
+        location: Optional[str] = None,
+        experiment: Optional[str] = None,
+        staging_bucket: Optional[str] = None,
+        credentials: Optional[auth_credentials.Credentials] = None,
+    ):
+        """Updates common initalization parameters with provided options.
+
+        Args:
+            project: The default project to use when making API calls.
+            location: The default location to use when making API calls. If not set
+                defaults to us-central-1
+            experiment: The experiment to assign
+            staging_bucket: The default staging bucket to use to stage artifacts when
+                making API calls.
+            credentials: The default custom credentials to use when making API calls. If
+                not provided crendentials will be ascertained from the environment.
+        """
+        if project:
+            self._project = project
+        if location:
+            utils.validate_region(location)
+            self._location = location
+        if experiment:
+            logging.warning("Experiments currently not supported.")
+            self._experiment = experiment
+        if staging_bucket:
+            self._staging_bucket = staging_bucket
+        if credentials:
+            self._credentials = credentials
+
+    @property
+    def project(self) -> str:
+        """Default project."""
+        if self._project:
+            return self._project
+
+        _, project_id = google.auth.default()
+        return project_id
+
+    @property
+    def location(self) -> str:
+        """Default location."""
+        return self._location if self._location else utils.DEFAULT_REGION
+
+    @property
+    def experiment(self) -> Optional[str]:
+        """Default experiment, if provided."""
+        return self._experiment
+
+    @property
+    def staging_bucket(self) -> Optional[str]:
+        """Default staging bucket, if provided."""
+        return self._staging_bucket
+
+    @property
+    def credentials(self) -> Optional[auth_credentials.Credentials]:
+        """Default credentials, if provided."""
+        return self._credentials
+
+    def get_client_options(
```

Comment:
In the currently generated surface `api_endpoint` is the only relevant client option but there are others (mostly auth related).

https://googleapis.dev/python/google-api-core/latest/client_options.html

How is this expected to be used? Is the idea that someone would someone call this, add additional client options, and then pass it to the client?

If you're constructing a client options for the user I'd prefer you create the object (for better type checking).

```py
from google.api_core.client_options import ClientOptions

options = ClientOptions(api_endpoint="foo.googleapis.com")
```

Reply:
This is great, `get_client_options()` is a helper method for us to construct the GAPIC client under-the-hood for the user. We should use the `api_core` object

Reply:
This method will be used by the _developers_ of Model Builder SDK when instantiating a GAPIC client. We will create GAPIC clients under the hood repeatedly throughout the MB SDK Classes and need to be able to use different endpoints based on the region or client type for every method call.

Here is a simple example:

**User 1 Code**
```
from google.cloud import aiplatform as aip

my_dataset = aip.Dataset(
  dataset_name="projects/123456/locations/europe-west4/datasets/987654"
)
```

**User 2 Code**
```
from google.cloud import aiplatform as aip

aip.init(location="asia-east1")
my_dataset = aip.Dataset(dataset_name="34627", location="us-central1")
```


**MB SDK Source Code**
```
from google.cloud.aiplatform import gapic
from google.cloud.aiplatform.utils import get_client_options

class Dataset:
    def __init__(self, dataset_name: str, location: Optional[str] = None):
        """Retrieves dataset resource."""

        self._client = gapic.DatasetServiceClient(
            client_options=get_client_options(location_override=location)
		)

        # 
        # More logic here
        #

        self._dataset = self._client.get_dataset(name=dataset_name)
```

Reply:
Thank you for the explanation @vinnysenthil.  I guess `client_options.api_endpoint` doesn't make much sense if the library is going to be using different endpoints. I still think there needs to be *some way* to pass user defined client options to the generated clients, maybe with the caveat that some will be ignored...

Reply:
Captured in this ticket [b/170489302](b/170489302)